File: Binary Trees and Binary Search Trees/700. Search in a Binary Search Tree/700. Search in a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        def search(root):
            if not root:
                return
                
            if root.val == val:
                return root
            elif root.val < val:
                return search(root.right)
            else:
                return search(root.left)

        return search(root)

# Time: O(n)
# Space: O(h)

File: Binary Trees and Binary Search Trees/700. Search in a Binary Search Tree/700. Search in a Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
var searchBST = function (root, val) {
  const search = (root) => {
    if (!root) return null;

    if (root.val === val) return root;
    else if (root.val < val) return search(root.right);
    else return search(root.left);
  };

  return search(root);
};

// Time: O(n)
// Space: O(n)
